axum/src/docs/routing/with_state.md

```diff
@@ -105,6 +105,14 @@ receives. That means it is not suitable for holding state derived from a
 request, such as authorization data extracted in a middleware. Use [`Extension`]
 instead for such data.
 
+# Mutable state
+
+As state is used across routes it is not mutable by default. To mutate a state
+you will need to use an `Arc<Mutex>` or similar. Accessing mutable IO such as 
+a database connection requires an async mutex or similar. See the 
+[tokio discussion on sync vs async mutex](https://docs.rs/tokio/1.25.0/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use). 
```

Comment:
This makes it sounds like sharing a database connection behind a mutex is common, which I don't think it is. Normally you share a connection pool which deals with all this internally, so users don't need a mutex.

You generally don't wanna share IO objects with mutexes, its a deadlock waiting to happen. In that case you wanna use [an actor](https://ryhl.io/blog/actors-with-tokio).

So I think we should phrase this differently somehow.